File: 프로그래머스/lv2/42883. 큰 수 만들기/큰 수 만들기.cpp
#include <string>
#include <vector>

using namespace std;

string solution(string number, int k) {
    string answer = "";
	int idx = 0;
	while(k>0 && idx<number.size())
	{
		if(number[idx] < number[idx+1])
		{
			number.erase(number.begin()+idx);
			idx = 0;
			k--;
		}
        else
        {
            if(idx == number.size()-2 && number[idx] >= number[idx+1])
            {
                number.erase(number.begin()+idx+1);
			    idx = 0;
			    k--;
                continue;
            }
		    idx++;
        }
	}
    return number;
}

File: 프로그래머스/lv2/42883. 큰 수 만들기/큰 수 만들기.java
import java.util.*;

class Solution {
    public String solution(String number, int k) {
        Stack<Character> stack = new Stack<>();
        char[] result = new char[number.length() - k];
        
        for(int i = 0; i < number.length(); i++){
            while(!stack.isEmpty() && stack.peek() < number.charAt(i) && k > 0){
                stack.pop();
                k--;
            }
            stack.push(number.charAt(i));
        }
        
        for(int i = 0; i < result.length; i++){
            result[i] = stack.get(i);
        }
        return new String(result);
    }
}
